### DISCUSSION TITLE
FutureProvider stuck on loading when called in method

### DISCUSSION THREAD

**Discussion by khari998** [rrousselGit/riverpod#1724] [Q&A]
I am rewriting an application in Riverpod and I am trying to use a FutureProvider in an onTap handler function so that I can load a dialog in another part of the application. 

I don't know if I'm using FutureProvider incorrectly but I haven't found a solution to solve this issue. 

I also don't know if this issue would provide a solution: https://github.com/rrousselGit/riverpod/issues/1660, if it is, please provide guidance as to how to implement this pattern properly.

Because the dialog is in a different widget, I am using a StateProvider to control a boolean to determine whether to show it or not that is dependent on the button tap

Sample application:

```dart
import` 'dart:convert';

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:http/http.dart' as http;

final exampleNetworkCallProvider =
    FutureProvider<Map<String, dynamic>>((ref) async {
  final exampleTodoResponse =
      await http.get(Uri.parse("https://jsonplaceholder.typicode.com/todos/4"));
  final todo = json.decode(exampleTodoResponse.body);
  print(todo); // Prints todo successfully
  return todo;
});

final isLoadingProvider = StateProvider((ref) => false);

void main() {
  runApp(const ProviderScope(child: MyApp()));
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Error Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'FutureBuilder Error'),
    );
  }
}

class MyHomePage extends ConsumerStatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  ConsumerState<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends ConsumerState<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    final bool loading = ref.watch(isLoadingProvider);
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            TextButton(
              onPressed: () {
               // Not working in an onPressed
                ref.watch(exampleNetworkCallProvider).when(
                      data: ((data) {
                        // should stop CircularProgressIndicator from rendering
                        ref
                            .watch(isLoadingProvider.notifier)
                            .update((state) => false);

                        // Data never arrives -- instead stuck in loading
                        print("data has arrived, $data");

                        // Reason accessing data from FutureProvider in onPressed method
                        if (data["completed"] as bool) {
                          // Load dialog
                        }
                      }),
                      error: ((error, stackTrace) => print("error ")),
                      loading: (() => ref
                          .watch(isLoadingProvider.notifier)
                          .update((state) => true)),
                    );
              },
              child: const Text("Press"),
            ),
            loading
                ? const CircularProgressIndicator()
                : const SizedBox.shrink()
          ],
        ),
      ),
    );
  }
}
```

**Comment by rrousselGit** [comment]
You're misunderstanding what `.when`does.
It's not listening to the result. It's a switch-case

You probably want to do:

```dart
onPressed: () async {
  final data = await ref.read(exampleNetworkCallProvider.future);
  print('data has arrived');
}
```

**Reply by khari998** [reply]
> You're misunderstanding what `.when`does. It's not listening to the result. It's a switch-case
> 
> You probably want to do:
> 
> ```dart
> onPressed: () async {
>   final data = await ref.read(exampleNetworkCallProvider.future);
>   print('data has arrived');
> }
> ```

So I thought about this in order to get the data from the API. But I still need access to the loading state in order to show the `CircularProgressIndicator` in another part of the application. 

How would you do that without using `.when()` or any other switch case method when you're only calling the FutureProvider in the `onPressed` method?

**Reply by rrousselGit** [reply]
Instead of doing ref.watch + when inside `onPressed`, you might want to do that inside the build method but not the onPressed

Like:
```dart
Widget build(context, ref) {
  final data = ref.watch(provider);
  return data.when(
    loading: () => CircularProgressIndicator(),
    error: (err, stack) => Text('error'),
    data: (data) => Text(data),
  );
}
```

**Reply by khari998** [reply]
> Instead of doing ref.watch + when inside `onPressed`, you might want to do that inside the build method but not the onPressed
> 
> Like:
> 
> ```dart
> Widget build(context, ref) {
>   final data = ref.watch(provider);
>   return data.when(
>     loading: () => CircularProgressIndicator(),
>     error: (err, stack) => Text('error'),
>     data: (data) => Text(data),
>   );
> }
> ```

I tried this too. The main issue with this approach is when the `CircularProgressIndicator()` needs to be displayed in another widget separate from where the `onPressed()` happens.  I'll use my real application flow as an example so that it is less confusing.

My application contains a survey. At the end of the survey is a button to submit the json data back to the server. While that json data is being submitted, I load a dialog box that shows the `CircularProgressIndicator()` until the response from the server comes back as a successful submission. 

I can't use the above approach because the FutureBuilder wouldn't trigger when the button is tapped to submit the survey data; It would instead trigger right when the button is rendered. Additionally, I can't avoid triggering the call to the server until after the survey is completed, so any call to the server has to originate from the `onPressed()` method. That leaves me with trying to call the call to the server with no way to access the loading state in order to update the `CircularProgressIndicator()` in the dialog box.

Also, I  really appreciate you taking the time to help me with this. I haven't been able to find any solution to this problem using Riverpod yet at all.

**Reply by rrousselGit** [reply]
It sounds like you're looking for https://github.com/rrousselGit/riverpod/issues/1660

**Reply by khari998** [reply]
> It sounds like you're looking for #1660
Yes I mentioned that earlier but am trying to piece together how it would work.

So theoretically I would just use the mutation in another component. To follow your example:  

`GetTodoMutation getTodo = ref.watch(exampleNetworkCallProvider.getTodo);` would be in the dialog where I can render the `CircularProgressIndicator()` And leave the network call in the onClicked? I believe this would solve the problem. I just wanted to make sure that I would be able to access the same loading state from the network call in a separate widget but you mentioned it should have that. 

Can't wait to see it released soon! Would really love to integrate this asap. Very fortunate that you were working on solving this problem right as I need to use it haha. Thanks for the hard work.
